File: packages/loopring_v3/operator/float.py
from collections import namedtuple

FloatEncoding = namedtuple("FloatEncoding", "numBitsExponent numBitsMantissa exponentBase")
Float28Encoding = FloatEncoding(5, 23, 10)
Float24Encoding = FloatEncoding(5, 19, 10)
Float16Encoding = FloatEncoding(5, 11, 10)
Float12Encoding = FloatEncoding(5,  7, 10)
Float8Encoding = FloatEncoding(5,  3, 10)

def toFloat(value, floatEncoding):
    maxPower = (1 << floatEncoding.numBitsExponent) - 1
    maxMantissa = (1 << floatEncoding.numBitsMantissa) - 1
    maxExponent = floatEncoding.exponentBase ** maxPower
    maxValue = maxMantissa * maxExponent
    assert value <= maxValue, "value too large: "

    exponent = 0
    r = value // maxMantissa
    d = 1
    while r >= floatEncoding.exponentBase or d * maxMantissa < value:
        r = r // floatEncoding.exponentBase
        exponent += 1
        d *= floatEncoding.exponentBase
    mantissa = value // d

    assert exponent <= maxExponent, "Exponent too large"
    assert mantissa <= maxMantissa, "Mantissa too large"
    f = (exponent << floatEncoding.numBitsMantissa) + mantissa
    return f

def fromFloat(f, floatEncoding):
    exponent = f >> floatEncoding.numBitsMantissa
    mantissa = f & ((1 << floatEncoding.numBitsMantissa) - 1)
    value = mantissa * (floatEncoding.exponentBase ** exponent)
    return value

def roundToFloatValue(value, encoding):
  f = toFloat(int(value), encoding)
  floatValue = fromFloat(f, encoding)
  return floatValue

File: packages/loopring_v3.js/src/lib/sign/float.ts
import BN from "bn.js";

export interface FloatEncoding {
  numBitsExponent: number;
  numBitsMantissa: number;
  exponentBase: number;
}

export function toFloat(value: BN, encoding: FloatEncoding) {
  const maxExponent = (1 << encoding.numBitsExponent) - 1;
  const maxMantissa = (1 << encoding.numBitsMantissa) - 1;
  const maxExponentValue = new BN(encoding.exponentBase).pow(
    new BN(maxExponent)
  );
  const maxValue = new BN(maxMantissa).mul(maxExponentValue);
  assert(
    value.lte(maxValue),
    "Value too large: " +
      value.toString(10) +
      ", max value: " +
      maxValue.toString(10)
  );

  let exponent = 0;
  let r = value.div(new BN(maxMantissa));
  let d = new BN(1);
  while (
    r.gte(new BN(encoding.exponentBase)) ||
    d.mul(new BN(maxMantissa)).lt(value)
  ) {
    r = r.div(new BN(encoding.exponentBase));
    exponent += 1;
    d = d.mul(new BN(encoding.exponentBase));
  }
  const mantissa = value.div(d).toNumber();

  assert(exponent <= maxExponent, "Exponent too large");
  assert(mantissa <= maxMantissa, "Mantissa too large");
  const f = (exponent << encoding.numBitsMantissa) + mantissa;
  return f;
}

export function fromFloat(f: number, encoding: FloatEncoding) {
  const exponent = f >> encoding.numBitsMantissa;
  const mantissa = f & ((1 << encoding.numBitsMantissa) - 1);
  const value = new BN(mantissa).mul(
    new BN(encoding.exponentBase).pow(new BN(exponent))
  );
  return value;
}

export function roundToFloatValue(value: BN, encoding: FloatEncoding) {
  const f = toFloat(value, encoding);
  const floatValue = fromFloat(f, encoding);
  assert(
    floatValue.lte(value),
    "float value can never be higher than the original value"
  );
  return floatValue;
}
